File: test/fixtures/php-project/src/Service/AuthService.php
<?php

namespace App\Services;

use App\Models\User;

/**
 * Contract for authentication services.
 */
interface AuthServiceInterface
{
    public function login(string $email, string $password): ?string;
    public function logout(string $token): void;
    public function validate(string $token): ?User;
}

/**
 * Default implementation of the authentication service.
 */
class AuthService implements AuthServiceInterface
{
    private array $sessions = [];

    /**
     * Authenticate a user by email and password.
     * Returns a session token on success, or null on failure.
     */
    public function login(string $email, string $password): ?string
    {
        // Simplified: accept any non-empty credentials
        if (empty($email) || empty($password)) {
            return null;
        }
        $token = bin2hex(random_bytes(16));
        $this->sessions[$token] = $email;
        return $token;
    }

    /**
     * Invalidate a session token.
     */
    public function logout(string $token): void
    {
        unset($this->sessions[$token]);
    }

    /**
     * Return the User associated with a token, or null if invalid.
     */
    public function validate(string $token): ?User
    {
        $email = $this->sessions[$token] ?? null;
        if ($email === null) {
            return null;
        }
        return new User(1, 'Auth User', $email);
    }

    private function hashPassword(string $password): string
    {
        return password_hash($password, PASSWORD_BCRYPT);
    }
}

File: test/fixtures/ruby-project/lib/auth_service.rb
require 'securerandom'
require_relative './models/user'

# Handles authentication: login, logout, token validation.
class AuthService
  # Maximum number of concurrent sessions allowed per user.
  MAX_SESSIONS = 10

  def initialize
    @sessions = {}
  end

  # Authenticate a user by email and password.
  # Returns a session token on success, or nil on failure.
  def login(email, password)
    return nil if email.nil? || password.nil?

    token = SecureRandom.hex(16)
    @sessions[token] = email
    token
  end

  # Invalidate an active session token.
  def logout(token)
    @sessions.delete(token)
  end

  # Validate a session token and return the associated email, or nil.
  def validate(token)
    @sessions[token]
  end

  # Returns the count of active sessions.
  def session_count
    @sessions.size
  end

  # Clear all sessions (admin operation).
  def self.clear_all(instance)
    instance.instance_variable_set(:@sessions, {})
  end

  # Create an AuthService with a pre-loaded set of sessions (test helper).
  def self.with_sessions(sessions)
    svc = new
    svc.instance_variable_set(:@sessions, sessions.dup)
    svc
  end

  private

  def expired?(token)
    false
  end
end
